worldedit-core/src/main/java/com/sk89q/worldedit/util/io/file/RecursiveDirectoryWatcher.java

```diff
@@ -0,0 +1,226 @@
+    private final WatchService watchService;
+    private Thread watchThread;
+    private Consumer<DirEntryChangeEvent> eventConsumer;
+    private BiMap<WatchKey, Path> watchRootMap = HashBiMap.create();
+
+    private RecursiveDirectoryWatcher(Path root, WatchService watchService) {
+        this.root = root;
+        this.watchService = watchService;
+    }
+
+    /**
+     * Create a new recursive directory watcher for the given root folder.
+     * You have to call @see start() before the instance starts monitoring.
+     *
+     * @param root Folder to watch for changed files recursively.
+     * @return A new RecursiveDirectoryWatcher instance, monitoring the given root folder.
+     * @throws IOException If creating the watcher failed, e.g. due to root not existing.
+     */
+    public static RecursiveDirectoryWatcher create(Path root) throws IOException {
+        WatchService watchService = root.getFileSystem().newWatchService();
+        return new RecursiveDirectoryWatcher(root, watchService);
+    }
+
+    private void registerFolderWatchAndScanInitially(Path root) throws IOException {
+        WatchKey watchKey = root.register(watchService, StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_DELETE);
+        LOGGER.debug("Watch registered: " + root);
+        watchRootMap.put(watchKey, root);
+        eventConsumer.accept(new DirectoryCreatedEvent(root));
+        for (Path path : Files.newDirectoryStream(root)) {
+            if (Files.isDirectory(path)) {
+                registerFolderWatchAndScanInitially(path);
+            } else {
+                eventConsumer.accept(new FileCreatedEvent(path));
+            }
+        }
+    }
+
+    /**
+     * Make this RecursiveDirectoryWatcher instance start monitoring the root folder it was created on.
+     * When this is called, RecursiveDirectoryWatcher will send initial notifications for the entire
+     * file structure in the configured root.
+     * @param eventConsumer The lambda that's fired for every file event.
+     */
+    public void start(Consumer<DirEntryChangeEvent> eventConsumer) {
+        this.eventConsumer = eventConsumer;
+        watchThread = new Thread(() -> {
+            LOGGER.debug("RecursiveDirectoryWatcher::EventConsumer started");
+
+            try {
+                registerFolderWatchAndScanInitially(root);
+            } catch (IOException e) { e.printStackTrace(); }
+
+            try {
+                WatchKey watchKey;
+                while (true) {
+                    try {
+                        watchKey = watchService.take();
+                    } catch (InterruptedException e) { break; }
+
+                    for (WatchEvent<?> event : watchKey.pollEvents()) {
+                        WatchEvent.Kind<?> kind = event.kind();
+                        if (kind.equals(StandardWatchEventKinds.OVERFLOW)) {
+                            LOGGER.warn("RecursiveDirectoryWatcher Seems like we can't keep up with updates");
```

Comment:
The class name is included in the logs.
```suggestion
                            LOGGER.warn("Seems like we can't keep up with updates");
```